pg_search/src/postgres/storage/fsm.rs

```diff
@@ -0,0 +1,149 @@
+use crate::postgres::storage::block::{BM25PageSpecialData, PgItem};
+use crate::postgres::storage::buffer::BufferManager;
+use pgrx::pg_sys;
+
+#[repr(transparent)]
+#[derive(Debug, Clone)]
+pub struct FreeBlockNumber(pg_sys::BlockNumber);
+
+impl From<pg_sys::BlockNumber> for FreeBlockNumber {
+    fn from(val: pg_sys::BlockNumber) -> Self {
+        FreeBlockNumber(val)
+    }
+}
+
+impl From<FreeBlockNumber> for pg_sys::BlockNumber {
+    fn from(val: FreeBlockNumber) -> Self {
+        val.0
+    }
+}
+
+impl From<FreeBlockNumber> for PgItem {
+    fn from(val: FreeBlockNumber) -> Self {
+        let bytes = val.0.to_ne_bytes();
+        let ptr = unsafe { pg_sys::palloc(bytes.len()) } as *mut i8;
+        unsafe { std::ptr::copy_nonoverlapping(bytes.as_ptr() as *const i8, ptr, bytes.len()) };
+        PgItem(
+            ptr as pg_sys::Item,
+            std::mem::size_of::<pg_sys::BlockNumber>(),
+        )
+    }
+}
+
+impl From<PgItem> for FreeBlockNumber {
+    fn from(pg_item: PgItem) -> Self {
+        let bytes = unsafe {
+            std::slice::from_raw_parts(
+                pg_item.0 as *const u8,
+                std::mem::size_of::<pg_sys::BlockNumber>(),
+            )
+        };
+        FreeBlockNumber(pg_sys::BlockNumber::from_ne_bytes(
+            bytes.try_into().unwrap(),
+        ))
+    }
+}
+
+pub struct FreeBlockList {
+    start_block_number: pg_sys::BlockNumber,
+    bman: BufferManager,
+}
+
+impl FreeBlockList {
+    ///
+    /// Open a new [`FreeBlockList`].
+    ///
+    /// # Arguments
+    ///
+    /// * `relation_oid` - The OID of the relation to vacuum.
+    /// * `start_block_number` - The block number of the first block in the list.
+    pub fn open(relation_oid: pg_sys::Oid, start_block_number: pg_sys::BlockNumber) -> Self {
+        let bman = BufferManager::new(relation_oid);
+        Self {
+            start_block_number,
+            bman,
+        }
+    }
+
+    ///
+    /// Pop a free block number from end of the list
+    ///
+    /// # Returns
+    ///
+    /// A [`FreeBlockNumber`] if one is available, or `None` if the list is empty.
+    pub fn pop(&mut self) -> Option<FreeBlockNumber> {
+        // Go to the end of the list, creating a list of block numbers along the way
+        let mut blockno = self.start_block_number;
+        let mut buffer = self.bman.get_buffer(blockno);
+        let mut blocknos = Vec::new();
+
+        while blockno != pg_sys::InvalidBlockNumber {
+            if blockno != self.start_block_number {
+                buffer = self.bman.get_buffer_exchange(blockno, buffer);
+            }
+
+            blocknos.push(blockno);
+            let page = buffer.page();
+            blockno = page.next_blockno();
+        }
+
+        drop(buffer);
```

Comment:
I think we should instead keep a `last_block_number: pg_sys::BlockNumber` that's updated whenever a block is pushed or popped onto/from the FreeBlockList.